api/subjects/serializers.py

```diff
@@ -0,0 +1,116 @@
+from rest_framework import serializers as ser
+from rest_framework import exceptions
+
+from framework.auth.core import Auth
+from framework.exceptions import PermissionsError
+
+from api.base.serializers import (
+    JSONAPISerializer,
+    JSONAPIRelationshipSerializer,
+    LinksField,
+    RelationshipField,
+    BaseAPISerializer,
+)
+from osf.exceptions import NodeStateError, ValidationValueError
+
+
+class UpdateSubjectsMixin(object):
+    def update_subjects_method(self, resource, subjects, auth):
+        # Method to update subjects on resource
+        raise NotImplementedError()
+
+    def update_subjects(self, resource, subjects, auth):
+        """Updates subjects on resource and handles errors.
+
+        :param object resource: Object for which you want to update subjects
+        :param list subjects: Subjects array (or array of arrays)
+        :param object Auth object
+        """
+        try:
+            self.update_subjects_method(resource, subjects, auth)
+        except PermissionsError as e:
+            raise exceptions.PermissionDenied(detail=str(e))
+        except ValueError as e:
+            raise exceptions.ValidationError(detail=str(e))
+        except ValidationValueError as e:
+            raise exceptions.ValidationError(detail=e[0])
+        except NodeStateError as e:
+            raise exceptions.ValidationError(detail=str(e))
+
+
+class SubjectSerializer(JSONAPISerializer):
+    filterable_fields = frozenset([
+        'text',
+        'parent',
+        'id',
+    ])
+    id = ser.CharField(source='_id', required=True)
+    text = ser.CharField(max_length=200)
+
+    parent = RelationshipField(
+        related_view='subjects:subject-detail',
+        related_view_kwargs={'subject_id': '<parent._id>'},
+    )
+
+    children = RelationshipField(
+        related_view='subjects:subject-children',
+        related_view_kwargs={'subject_id': '<_id>'},
+        related_meta={'count': 'get_children_count'},
+
+    )
+
+    links = LinksField({
+        'self': 'get_absolute_url',
+    })
+
+    def get_absolute_url(self, obj):
+        return obj.absolute_api_v2_subject_url
+
+    def get_children_count(self, obj):
+        return obj.child_count()
+
+    class Meta:
+        type_ = 'subjects'
+
+
+class SubjectRelated(JSONAPIRelationshipSerializer):
+    id = ser.CharField(source='_id', required=False, allow_null=True)
+    class Meta:
+        type_ = 'subjects'
+
+
+class SubjectsRelationshipSerializer(BaseAPISerializer, UpdateSubjectsMixin):
```

Comment:
Serializes a resources subjects' self link:
GET http://localhost:8000/v2/nodes/9m4wp/relationships/subjects/
```
{
    "data": [
        {
            "type": "subjects",
            "id": "5c92ca721dbe200001acaa92"
        },
        {
            "type": "subjects",
            "id": "5c92ca841dbe200001acaec9"
        }
    ],
    "links": {
        "self": "http://localhost:8000/v2/nodes/9m4wp/relationships/subjects/",
        "html": "http://localhost:8000/v2/nodes/9m4wp/subjects/"
    },
    "meta": {
        "version": "2.15"
    }
}
```

While you can update subjects with this serializer, you can usually just update through the resource itself. Main purpose is to remove subjects entirely.